[Merged by Bors] - Standard Material Blend Modes

coreh opened the pull request:
# Objective

- This PR adds support for blend modes to the PBR `StandardMaterial`.

<img width="1392" alt="Screenshot 2022-11-18 at 20 00 56" src="https://user-images.githubusercontent.com/418473/202820627-0636219a-a1e5-437a-b08b-b08c6856bf9c.png">

<img width="1392" alt="Screenshot 2022-11-18 at 20 01 01" src="https://user-images.githubusercontent.com/418473/202820615-c8d43301-9a57-49c4-bd21-4ae343c3e9ec.png">

## Solution

- The existing `AlphaMode` enum is extended, adding three more modes: `AlphaMode::Premultiplied`, `AlphaMode::Add` and `AlphaMode::Multiply`;
- All new modes are rendered in the existing `Transparent3d` phase;
- The existing mesh flags for alpha mode are reorganized for a more compact/efficient representation, and new values are added;
- `MeshPipelineKey::TRANSPARENT_MAIN_PASS` is refactored into `MeshPipelineKey::BLEND_BITS`.
  -  `AlphaMode::Opaque` and `AlphaMode::Mask(f32)` share a single opaque pipeline key: `MeshPipelineKey::BLEND_OPAQUE`;
  - `Blend`, `Premultiplied` and `Add` share a single premultiplied alpha pipeline key, `MeshPipelineKey::BLEND_PREMULTIPLIED_ALPHA`. In the shader, color values are premultiplied accordingly (or not) depending on the blend mode to produce the three different results after PBR/tone mapping/dithering;
  - `Multiply` uses its own independent pipeline key, `MeshPipelineKey::BLEND_MULTIPLY`;
- Example and documentation are provided.
---

## Changelog

### Added

- Added support for additive and multiplicative blend modes in the PBR `StandardMaterial`, via `AlphaMode::Add` and `AlphaMode::Multiply`;
- Added support for premultiplied alpha in the PBR `StandardMaterial`, via `AlphaMode::Premultiplied`;

alice-i-cecile reviewed (changes requested):
Fantastic stuff: great docs, solid code quality and a very useful feature.

The example is great, but would be *much* more useful for learning with labels under each sphere that record which blend mode is being used.

StarArawn commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -268,3 +269,16 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
+fn premultiply_alpha(standard_material_flags: u32, color: vec4<f32>) -> vec4<f32> {
+    let alpha_mode = standard_material_flags & STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BITS;
+    if (alpha_mode == STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BLEND) {
+        return vec4<f32>(color.rgb * color.a, color.a);
```
Is this correct? I'm not sure standard blending should multiple the color by the alpha. 🤔

coreh replied:
It's counterintuitive, but I believe it is correct:

As suggested by (I believe?) @fintelia, to avoid the inefficiency of having three separate pipelines, the `Blend`, `Premultiplied` and `Add` modes all share the same pipeline key (`MeshPipelineKey::BLEND_PREMULTIPLIED_ALPHA`) which produces the same blend state (`BlendState::PREMULTIPLIED_ALPHA_BLENDING`) — changed from `BlendState::ALPHA_BLENDING` that we used before.

This mode assumes alpha is premultiplied, and depending on what we do with the alpha channel and the color we get different results:

- If we multiply the color and keep the alpha channel, we restore the previous behavior we had under `BlendState::ALPHA_BLENDING` (used for `Blend`)
- If we multiply the color but change the alpha channel to `0.0`, we get additive blending
- If we don't change the color at all, we get the normal/expected behavior of `BlendState::PREMULTIPLIED_ALPHA_BLENDING`.

coreh replied:
Added some comments clarifying the math to the shader code.

IceSentry commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -101,5 +101,8 @@ fn fragment(in: FragmentInput) -> @location(0) vec4<f32> {
         output_color = dither(output_color, in.frag_coord.xy);
 #endif
+
+    output_color = premultiply_alpha(material.flags, output_color);
```
I feel like this should be behind a shader def. It's probably not that expensive, but still, when the mesh is opaque it doesn't need to be in the shader.

coreh replied:
Moved it to behind a shader def

IceSentry commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -268,3 +269,16 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
 
+fn premultiply_alpha(standard_material_flags: u32, color: vec4<f32>) -> vec4<f32> {
+    let alpha_mode = standard_material_flags & STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BITS;
+    if (alpha_mode == STANDARD_MATERIAL_FLAGS_ALPHA_MODE_BLEND) {
```
would using a switch statement be better here? That's a lot of conditional checks it would need to do to get to the last branch and I have no idea if the compiler is smart enough to optimize it.

coreh replied:
Looks like switches in WGSL only allow literals:

<img width="801" alt="image" src="https://user-images.githubusercontent.com/418473/202330992-92ca87d2-a37a-444d-99bb-fb135130d198.png">

I could add comments (or maybe defines instead of constants?) but maybe it would be less readable? Let me know what you prefer

coreh replied:
Using an additional shader def, I cleaned it up so that only two conditionals are needed now. Hopefully it will optimize better.

coreh commented:
> The example is great, but would be much more useful for learning with labels under each sphere that record which blend mode is being used.

That makes sense, Is there a good way of displaying text on the 3d scene? Or do I create 2d text and just reposition it somehow based on the camera position/projection?

coreh commented on examples/3d/blend_modes.rs:
```diff
@@ -0,0 +1,257 @@
+        PbrBundle {
+            mesh: icosphere_mesh.clone(),
+            material: materials.add(StandardMaterial {
+                base_color,
```
I considered doing that originally, but ended up not doing it for two reasons:

- It's useful to understand how premultiplied fades between an add-like effect to a traditional alpha effect as you increase the opacity
- Changing the `base_color` is not enough because of ambient light and the specular highlight, so it still looks kinda weird.

Thinking about it, having three spheres that look exactly the same when the demo starts is _also_ confusing. If I understand it correctly, premultiplied is mostly useful for things like foliage, since you can get alpha without the ugly outline, so maybe a separate demo (comparing it to just blend, or maybe blend and mask) makes more sense?

I can try to come up with something, let me know what y'all think.

StarArawn replied:
> Changing the base_color is not enough because of ambient light and the specular highlight, so it still looks kinda weird.

Reading up on the subject I think the alpha needs to be applied to these possibly? As the alpha decreasing means a decrease in light contribution.

I think premultiplied alpha is quite useful for things outside of folliage. In fact nvidia makes mention of better results with premultiplied alpha vs traditional here:
https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre

coreh commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -268,3 +269,16 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
 }
 #endif
 
+fn premultiply_alpha(standard_material_flags: u32, color: vec4<f32>) -> vec4<f32> {
```
Added documentation.

coreh commented:
`Text2d` didn't work, but searching `#help` I was able to find some sample code for using `camera.world_to_viewport()` and UI text:

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/202350584-21ed34e2-f53d-4740-b479-1c6af13525aa.png">

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/202350652-fe2370c0-6af3-4fbc-8271-821eb0236871.png">

Also added a toggle to verify the blend modes work properly with HDR on. (They do!) The results are _slightly_ different (particularly in specular highlights) but I think the result with HDR looks a bit better overall:

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/202350674-b92920c5-d63a-4c00-844c-f0dbac0d5734.png">

JMS55 commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -295,9 +295,13 @@ bitflags::bitflags! {
-        const ALPHA_MODE_OPAQUE          = (1 << 6);
-        const ALPHA_MODE_MASK            = (1 << 7);
-        const ALPHA_MODE_BLEND           = (1 << 8);
+        const ALPHA_MODE_BITS            = (0b111 << 6);
```
Can we have a small comment explaining what this bitflag is for / how the bitflags for the alpha modes are arranged?

coreh replied:
I added a comment, and changed it to keep the binary literal just for the mask, using sequential decimals for the values to show there's no special bitwise magic going on (except for the mask). Also renamed the masks for consistency with other reserved bits masks we already had.

alice-i-cecile reviewed (approved):
I'm happy with this now :) I'd update the image in the PR description, to make it easier for people preparing the release notes to snag a good screenshot.

OriginLive commented:
Hi, what I'm trying to do is to get an inverted color image
![image](https://user-images.githubusercontent.com/24319526/202927337-25f9680a-3042-4f2e-8a84-bf82a4b9501c.png)
An example of a thread in unity: https://forum.unity.com/threads/invert-colors-shader.205244/

To do this, i need to insert different blend more fns, akin to this:


```rs
glEnable(GL_BLEND);
glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
```

Would it be possible to add the blend function?

OriginLive commented on crates/bevy_pbr/src/render/mesh.rs:
```diff
@@ -620,12 +623,30 @@ impl SpecializedMeshPipeline for MeshPipeline {
+            blend = Some(BlendState {
+                color: BlendComponent {
+                    src_factor: BlendFactor::Dst,
+                    dst_factor: BlendFactor::OneMinusSrcAlpha,
```
Can we expose the blend component to shader somehow in a custom blend mode?

coreh replied:
While exposing `BlendComponent`, `BlendFactor` and `BlendOperation` (like Unity seems to do) directly via the `StandardMaterial`  _could_ be done:

- To support the full range of combinations, it would require using at least 11 additional bits of the `MeshPipelineKey`; (Instead of just two bits for the current approach) — and up to 22 bits if you wanted to also support the alpha options.
- Most of the possible combinations are either meaningless or not very useful when combined with the PBR `StandardMaterial`.

So I'm not sure if it makes sense, but would like others to chime in.

One thing we _could do_ (but ideally as a separate PR) is to introduce a `fn blend_state() -> Option<BlendState>` method to the `Material` trait, so that we don't have to use the [somewhat large specialization code snippet](https://github.com/bevyengine/bevy/pull/6644#issuecomment-1321413203) I left on my other comment.

OriginLive replied:
It would be nice to have an option to use it, rather than gate it arbitrarily, in my case I needed it.

I would much rather have the user pick what options they need than have a few preselected ones

alice-i-cecile replied:
I think a custom material is simple enough that I would suggest users do that if they need more unusual behavior.

coreh commented:
@OriginLive Do you need the inverted region to be somehow modulated by a mesh that _also_ has PBR shading (i.e. lighting, shadows on the invert mesh somehow reduces/increases the invert effect of other meshes behind it)?

If all you need is a mesh that completely inverts whatever's behind it, you can achieve this today (in a bit verbose manner) via specialization:

```rust
use bevy::{
    reflect::TypeUuid,
    render::render_resource::{AsBindGroup, BlendComponent, BlendFactor, BlendState, ShaderRef},
};

#[derive(AsBindGroup, TypeUuid, Debug, Clone)]
#[uuid = "fbe2f06f-94c1-4801-b2f9-3f13c98d5a88"]
struct InvertMaterial {}

impl Material for InvertMaterial {
    fn specialize(
        _pipeline: &MaterialPipeline<Self>,
        descriptor: &mut RenderPipelineDescriptor,
        _layout: &bevy::render::mesh::MeshVertexBufferLayout,
        _key: MaterialPipelineKey<Self>,
    ) -> Result<(), SpecializedMeshPipelineError> {
        descriptor.fragment.as_mut().unwrap().targets[0]
            .as_mut()
            .unwrap()
            .blend = Some(BlendState {
            color: BlendComponent {
                src_factor: BlendFactor::OneMinusDst,
                dst_factor: BlendFactor::Zero,
                operation: BlendOperation::Add,
            },
            alpha: BlendComponent::OVER,
        });

        Ok(())
    }

    fn fragment_shader() -> ShaderRef {
        "shaders/white.wgsl".into()
    }

    fn alpha_mode(&self) -> AlphaMode {
        AlphaMode::Blend
    }
}
```

Then for the shader, in `assets/shaders/white.wgsl`:

```wgsl
@fragment
fn fragment(
    #import bevy_pbr::mesh_vertex_output
) -> @location(0) vec4<f32> {
    return vec4<f32>(1.0, 1.0, 1.0, 1.0);
}
```

I have a working example on a separate branch, built on top of this blend modes branch, for convenience, but you can copy and paste this material/shader and it should work on vanilla bevy 0.9, right now:

https://github.com/coreh/bevy/commit/9ea64a88d97b6b32cdc77cfe032084f0b7e51d3f

coreh commented:
@fintelia Thinking about it more, one downside of reusing `BlendState::PREMULTIPLIED_ALPHA_BLENDING` for `AlphaMode::Blend`, `AlphaMode::Premultiplied` and `AlphaMode::Add` is that it requires all custom shader materials using `Blend` to add an extra final step premultiplying their RGB channels, (similar to what we do in the `premultiply_alpha()` function, which is a breaking change I did not originally anticipate.

If we make them separate that will require 1 extra bit for the `MeshPipelineKey` but will not be a breaking change for those shaders.

How prevalent are materials with custom shaders using `AlphaMode::Blend`?

OriginLive commented:
> 

My use case is to create a cursor that changes it's color based on the environment it's looking at.
It would be great if it'd work with pbr for other people, but that's good enough for my usecase (although it doesn't really work with hdr).

If possible, an example of blend usages would be nice.

IceSentry reviewed (approved):
The implementation looks good to me, the comments in the code are really nice.

The example is a bit bigger than I would like for an example, but it's not a blocker for me since it shows the feature really well.

alice-i-cecile commented:
Holding off for one more rendering approval: I want to be sure that the math and framing is correct.

IceSentry commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -268,3 +269,66 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
+        //
+        // Which is the blend operation for regular alpha blending `BlendState::ALPHA_BLENDING`
+        return vec4<f32>(color.rgb * color.a, color.a);
+    } if (alpha_mode == STANDARD_MATERIAL_FLAGS_ALPHA_MODE_ADD) {
```
Oh yeah, good catch, it does seems like it's missing an else
```suggestion
    } else if (alpha_mode == STANDARD_MATERIAL_FLAGS_ALPHA_MODE_ADD) {
```

coreh replied:
Nice catch. The lack of something like `rustfmt` for `wgsl` really makes me let a bunch of weird formatting errors slip by 😅

JMS55 commented on crates/bevy_pbr/src/alpha.rs:
```diff
@@ -23,6 +23,30 @@ pub enum AlphaMode {
+    /// Similar to [`AlphaMode::Blend`], however assumes RGB channel values are
+    /// [premultiplied](https://en.wikipedia.org/wiki/Alpha_compositing#Straight_versus_premultiplied).
+    ///
+    /// For fixed RGB values, behaves more like [`AlphaMode::Blend`] for
```
What do you mean by fixed RGB values?

coreh replied:
Hmm yeah that came out confusing. Changed it to “For otherwise constant RGB values...”

robtfm commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -265,3 +266,66 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
+        //     result = src_alpha * src_color + (1 - src_alpha) * dst_color
+        //
+        // Which is the blend operation for regular alpha blending `BlendState::ALPHA_BLENDING`
+        return vec4<f32>(color.rgb * color.a, color.a);
```
do we risk losing precision here, compared to returning unmultiplied and using `BlendState::ALPHA_BLENDING`? i think probably not but i'm not 100% sure. 

I don't really see a reason to change this path though, we could just not push `PREMULTIPLY_ALPHA` for `AlphaMode::Blend` and set the blend state as before. it would save a few registers and preserve precision. arguably it's conceptually simpler as well?

robtfm replied:
the nvidia paper https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre from comments above gives a good reason for the change: it behaves much better with mipmapping.

coreh replied:
It's been a while since I wrote this PR; Consolidating additive, premultiplied and regular alpha blending into a single pass was based off some early feedback. I don't recall the _exact_ reasoning, but looking at the code, and if I understand it correctly, the benefits I see are:

- We keep blend mode reserved bits to 2 in the mesh pipeline key, (instead of 3) so we get to keep more space for future needs there.
- Blended, Premultiplied and Add meshes are all rendered in a single go (?) so we avoid changing the pipeline. I don't understand WGPU enough to know if this is a performance win or not, my hunch is that... it depends?

coreh replied:
> the nvidia paper https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre from comments above gives a good reason for the change: it behaves much better with mipmapping.

That was entirely coincidental, at least on my part! 😆 I'm trying to find who originally suggested consolidating, maybe they had that in mind as well?

superdump replied:
Being able to use the same pipeline for more things is good as it allows us to batch multiple draw commands into one.

robtfm reviewed (commented):
this is awesome

superdump reviewed (changes requested):
Please rebase. And there are just a couple of minor adjustments to make then we can merge this. Thanks and thanks for your patience! :)

coreh commented on crates/bevy_pbr/src/material.rs:
```diff
@@ -386,8 +386,14 @@ pub fn queue_material_meshes<M: Material>(
                         let alpha_mode = material.properties.alpha_mode;
-                        if let AlphaMode::Blend = alpha_mode {
-                            mesh_key |= MeshPipelineKey::TRANSPARENT_MAIN_PASS;
+                        if let AlphaMode::Blend | AlphaMode::Premultiplied | AlphaMode::Add =
+                            alpha_mode
```
Yeah, just double checked it, and it does work as expected 

[Playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=95bd23019392da1afb62dc61592a9098)

I didn't think much of it as I was making this change, since I just reused the existing `if let` statement, but looking back it does look kinda cleaner than `if alpha_mode == AlphaMode::Blend || alpha_mode == AlphaMode::Premultiplied || alpha_mode == AlphaMode::Add` 😄

superdump replied:
I agree it’s more concise. I think the mental hiccup for me is because the flags can be combined using the | or operator. But I don’t mind either way. This is fine.

coreh commented on crates/bevy_pbr/src/render/mesh.rs:
```diff
@@ -622,12 +625,30 @@ impl SpecializedMeshPipeline for MeshPipeline {
             // but their depth is not written to the depth buffer
             depth_write_enabled = false;
+        } else if pass == MeshPipelineKey::BLEND_MULTIPLY {
+            label = "multiply_mesh_pipeline".into();
```
I _can_ change this, but it will be slightly inconsistent with the other two, that do match 1:1 with the key name, e.g.

- `MeshPipelineKey::PREMULTIPLY_ALPHA` → `premultiply_alpha_mesh_pipeline`
- `MeshPipelineKey::OPAQUE` → `opaque_mesh_pipeline`

Let me know what you prefer

superdump replied:
Maybe POSTMULTIPLY_ALPHA and postmultiply_alpha_mesh_pipeline?

superdump replied:
Stick with multiply.

coreh commented:
> Please rebase. And there are just a couple of minor adjustments to make then we can merge this.

Updated to the latest `main`!

> Thanks and thanks for your patience! :)

No problem 😄

Just moved all of the blend mode-related flags/keys to the upper portion, and made the “regular” flags sequential again.

I couldn't get const expressions to work in WGSL, (even though they're mentioned on [the spec](https://gpuweb.github.io/gpuweb/wgsl/#const-expr) — maybe the compiler is 100% compliant or up to date?) so we have some pretty ugly large numbers 😕

superdump commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -298,20 +298,25 @@ bitflags::bitflags! {
+        const TWO_COMPONENT_NORMAL_MAP   = (1 << 6);
+        const FLIP_NORMAL_MAP_Y          = (1 << 7);
+        const ALPHA_MODE_RESERVED_BITS   = (Self::ALPHA_MODE_MASK_BITS << Self::ALPHA_MODE_SHIFT_BITS); // ← Bitmask reserving bits for the `AlphaMode`
+        const ALPHA_MODE_OPAQUE          = (0 << Self::ALPHA_MODE_SHIFT_BITS);                          // ← Values are just sequential values bitshifted into
```
0 << n won’t work.

superdump commented on crates/bevy_pbr/src/render/mesh.rs:
```diff
@@ -557,13 +557,16 @@ bitflags::bitflags! {
+        const NORMAL_PREPASS              = (1 << 4);
+        const ALPHA_MASK                  = (1 << 5);
+        const BLEND_RESERVED_BITS         = Self::BLEND_MASK_BITS << Self::BLEND_SHIFT_BITS; // ← Bitmask reserving bits for the blend state
+        const BLEND_OPAQUE                = (0 << Self::BLEND_SHIFT_BITS);                   // ← Values are just sequential within the mask, and can range from 0 to 3
```
Same here

superdump commented on crates/bevy_pbr/src/render/pbr_types.wgsl:
```diff
@@ -17,15 +17,17 @@ let STANDARD_MATERIAL_FLAGS_METALLIC_ROUGHNESS_TEXTURE_BIT: u32 = 4u;
+let STANDARD_MATERIAL_FLAGS_TWO_COMPONENT_NORMAL_MAP: u32       = 64u;
+let STANDARD_MATERIAL_FLAGS_FLIP_NORMAL_MAP_Y: u32              = 128u;
+let STANDARD_MATERIAL_FLAGS_ALPHA_MODE_RESERVED_BITS: u32       = 3758096384u; // (0b111u32 << 29)
+let STANDARD_MATERIAL_FLAGS_ALPHA_MODE_OPAQUE: u32              = 0u;          // (0u32 << 29)
```
0 means no flags are set, you’ll have to start from 1 << 29.

superdump replied:
Oh… I see. You mean that none of the 3 bits are set. Ok.

superdump commented:
I think either start from 1<<29 or leave the 0 value as being an implicit fallback. I prefer starting from 1<<29 and being explicit if you feel that is ok?

coreh commented:
The `(0 << 29)` does look weird, haha. I left it there mostly so that it was visually consistent with the other ones, and aligned properly for the comments.

Yeah, these 3 (or 2) bits behave more like a bit-masked enum rather than fully independent flags, which is why the `0b000` state is the first of the values.  If you prefer, I _can_ make the change, and have the `1` be the first value, however:

- `MeshPipelineKey::NONE` and `StandardMaterialFlags::NONE` will be invalid, since the `BLEND_RESERVED_BITS` and `ALPHA_MODE_RESERVED_BITS` will be by default in a combination that doesn't match any valid blend mode; Right now the `NONE` states of both types conveniently map to `MeshPipelineKey::BLEND_OPAQUE` and `StandardMaterialFlags::ALPHA_MODE_OPAQUE`.
- In the future, if we ever want to add more values it will be slightly wasteful since one of the values will be explicitly unused.

superdump commented:
Mmm. Ok. I’m fine with leaving them as zero.

superdump commented:
bors r+
